modules/core/src/ocl.cpp

```diff
@@ -3078,7 +3078,7 @@ bool Kernel::run(int dims, size_t _globalsize[], size_t _localsize[],
             dims == 1 ? 64 : dims == 2 ? (i == 0 ? 256 : 8) : dims == 3 ? (8>>(int)(i>0)) : 1;
         CV_Assert( val > 0 );
         total *= _globalsize[i];
-        if (_globalsize[i] == 1)
+        if (_globalsize[i] == 1 && !_localsize)
```

Comment:
@dkurt What test case requires this change?

Reply:
@alalek, With gemm for small matrices.
https://github.com/opencv/opencv/blob/efceafcb5fa51142e14fc10b1e163e22888e8fa6/modules/core/test/ocl/test_gemm.cpp#L148-L161